core/src/Dscp/Core/Transactions/Types.hs

```diff
@@ -0,0 +1,276 @@
+
+-- This is all naive for now, should be moved to the separate module later
+
+-- | Coin amount.
+newtype Coin = Coin { unCoin :: Word64 }
+    deriving (Eq, Ord, Show, Generic, Hashable, Num)
+
+-- | Safely convert coin to integer.
+coinToInteger :: Coin -> Integer
+coinToInteger = toInteger . unCoin
+
+instance Buildable Coin where
+    build (Coin c) = c ||+ " coin(s)"
+
+----------------------------------------------------------------------------
+-- Transactions
+----------------------------------------------------------------------------
+
+-- We have several different types of "transactions". Money
+-- transactions, delegation transactions (in the future). Money
+-- transactions are the most popular, so we'll call them just
+-- "transactions".
+
+-- | Tx input account. Can be used for other tx types too.
+data TxInAcc = TxInAcc
+    { tiaAddr  :: Address
+    , tiaNonce :: Integer
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable TxInAcc where
+    build TxInAcc{..} = "TxInnAcc {" +| tiaAddr |+ " nonce " +|| tiaNonce ||+ "}"
+
+-- | Money transaction output.
+data TxOut = TxOut
+    { txOutAddr  :: Address
+    , txOutValue :: Coin
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable TxOut where
+    build TxOut{..} = "<" +| txOutAddr |+ ", " +| txOutValue |+ ">"
+
+-- | Transaction. Accounting-style money transfer.
+data Tx = Tx
+    { txInAcc   :: TxInAcc
+    , txInValue :: Coin
+    , txOuts    :: [TxOut]
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable Tx where
+    build Tx{..} =
+        "Tx { from: " +| txInAcc |+ "; inValue: " +| txInValue |+ "; outs:" +| listF txOuts |+ " }"
+
+type TxId = Hash Tx
+
+data PublicationTx = PublicationTx
+    { ptAuthor    :: Address
+    , ptPrevBlock :: Maybe (Hash PrivateBlock)
+    , ptBlock     :: Hash PrivateBlock
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable PublicationTx where
+    build PublicationTx { ptAuthor, ptPrevBlock, ptBlock } =
+        "Tx { from: " +| ptAuthor |+ "; prev block: " +| ptPrevBlock |+ "; block:" +| ptBlock |+ " }"
+
+type PublicationTxId = Hash PublicationTx
+
+-- | Compute tx id.
+txId :: (Serialise Tx) => Tx -> TxId
+txId = hash
+
+-- | Compute tx id.
+ptxId :: (Serialise PublicationTx) => PublicationTx -> PublicationTxId
+ptxId = hash
+
+-- | Transaction witness. We sign a pair of transaction hash and private
+-- key. The second element is there to authenticate proposed changes
+-- (@kirill.andreev). Public key hash should be equal to the input address.
+-- Also, public key should be the same which used to validate signature.
+data TxWitness = TxWitness
+    { txwSig :: Signature (TxId, PublicKey, ())
+    , txwPk  :: PublicKey
+    } deriving (Eq, Show, Generic)
+
+instance Buildable TxWitness where
+    build TxWitness {..} =
+        "TxWitness { " +| txwSig |+ ", pk: " +| txwPk |+ " }"
+
+-- | Incoming publication message; transaction payload.
+data Publication blockHash addr = Publication
```

Comment:
Wait, why did you go with `blockHash` parameter? Also, why is `addr` phantom?

Reply:
This type was born in the `PublicationValidator` and is required to have `blockHash` parameter to avoid adding `publication` type parameter to validator. Look onto the validator.